src/matrix_cookbook/for_mathlib/ring.inverse.lean

```diff
@@ -0,0 +1,85 @@
+import data.matrix.basic
+import linear_algebra.matrix.nonsingular_inverse
+
+variables {R: Type*}[division_ring R][group R]
```

Comment:
This is nonsense, it means "let `R` have two unrelated multiplication operators"!

Reply:
Yes you are right! Perhaps I should explain my conundrum better, but this will be long:

## Attempt 1

The following code Does not typecheck for `^-1` i.e. the inverse is not defined. 
```
import data.matrix.basic
import linear_algebra.matrix.nonsingular_inverse
variables {R: Type*}[ring R]
variables (A B C : R)
lemma eq_161 {hA: A ≠ 0}{hAadd1: A + 1 ≠ 0}: 
  (1 + A⁻¹)⁻¹ = A*(A + 1)⁻¹ := begin
  apply inv_eq_of_mul_eq_one_right,
  rw ← mul_assoc,
  rw [←  mul_assoc, add_mul, inv_mul_cancel hA, one_mul, mul_inv_cancel hAadd1],
end
```

```
failed to synthesize type class instance for
R : Type u_1,
_inst_1 : ring R,
A : R,
hA : A ≠ 0,
hAadd1 : A + 1 ≠ 0
⊢ has_inv R
```

## Attempt 2
So I try:

```
import data.matrix.basic
import linear_algebra.matrix.nonsingular_inverse

variables {R: Type*}[ring R]
variables (A B C : R)

lemma eq_161b {hA: A ≠ 0}{hAadd1: A + 1 ≠ 0}: 
  ring.inverse (1 + ring.inverse A) = A* (ring.inverse (A + 1)) := begin
  apply inv_eq_of_mul_eq_one_right,
  rw ← mul_assoc,
  rw [←  mul_assoc, add_mul, inv_mul_cancel hA, one_mul, mul_inv_cancel hAadd1],
end
``` 
which fails with 

```
invalid apply tactic, failed to unify
  ring.inverse (1 + ring.inverse A) = A * ring.inverse (A + 1)
with
  ?m_3⁻¹ = ?m_4
state:
R : Type u_1,
_inst_1 : ring R,
A : R,
hA : A ≠ 0,
hAadd1 : A + 1 ≠ 0
⊢ ring.inverse (1 + ring.inverse A) = A * ring.inverse (A + 1)
```
## Attempt 3

So then I say perhaps the operation `ring.inverse` is not the same as `inv`.  But I find a lemma `ring.inverse_eq_inv`, so I try to add that `rw ring.inverse_eq_inv A` but that does not work! I get the error message:
```
failed to synthesize type class instance for
R : Type u_1,
_inst_1 : ring R,
A : R,
hA : A ≠ 0,
hAadd1 : A + 1 ≠ 0
⊢ group_with_zero R
state:
R : Type u_1,
_inst_1 : ring R,
A : R,
hA : A ≠ 0,
hAadd1 : A + 1 ≠ 0
⊢ ring.inverse (1 + ring.inverse A) = A * ring.inverse (A + 1)
```

## Attempt 4
So then I say perhaps I am missing some instances. I try `field` in place of ring. Ofcourse it works! So I try to find the "minimum" that works. `ring`  alone does not work.
I check lemma `inv_eq_of_mul_eq_one_right`, turns out it needs `division_monoid` instance. So I add that. Well it does not work! 
.....
## Many random stuff in the middle.

## Attempt n
In the middle mul_assoc refuses to see an obvious term so I add the group term so .... more random attempts
After a while of random attempts I get to what you see above!!

This is turning into more of a rant so I will stop here. And end with a question?

## What instances should I start with on R!!!